sklearn/tree/_tree.pyx

```diff
@@ -867,6 +868,139 @@ cdef class Tree:
 
         return out
 
+    cpdef np.ndarray apply_depth(self, object X):
+        """Finds the depth level of the leaf node for each sample in X."""
+        if issparse(X):
+            return self._apply_depth_sparse_csr(X)
+        else:
+            return self._apply_depth_dense(X)
+
+    cdef inline np.ndarray _apply_depth_dense(self, object X):
+        """Finds the depth level of the leaf node for each sample in X."""
+
+        # Check input
+        if not isinstance(X, np.ndarray):
+            raise ValueError("X should be in np.ndarray format, got %s"
+                             % type(X))
+
+        if X.dtype != DTYPE:
+            raise ValueError("X.dtype should be np.float32, got %s" % X.dtype)
+
+        # Extract input
+        cdef np.ndarray X_ndarray = X
+        cdef DTYPE_t* X_ptr = <DTYPE_t*> X_ndarray.data
+        cdef SIZE_t X_sample_stride = <SIZE_t> X.strides[0] / <SIZE_t> X.itemsize
+        cdef SIZE_t X_fx_stride = <SIZE_t> X.strides[1] / <SIZE_t> X.itemsize
+        cdef SIZE_t n_samples = X.shape[0]
+
+        # Initialize output
+        cdef np.ndarray[DOUBLE_t] depth = np.zeros((n_samples,), dtype=np.float64)
+        cdef DOUBLE_t* depth_ptr = <DOUBLE_t*> depth.data
+
+
+        # Initialize auxiliary data-structure
+        cdef Node* node = NULL
+        cdef SIZE_t i = 0
+        cdef DOUBLE_t current_depth = 0.
+
+        with nogil:
+            for i in range(n_samples):
+                node = self.nodes
+                current_depth = 0.
+                # While node not a leaf
+                while node.left_child != _TREE_LEAF:
+                    # ... and node.right_child != _TREE_LEAF:
+                    if X_ptr[X_sample_stride * i +
+                             X_fx_stride * node.feature] <= node.threshold:
+                        node = &self.nodes[node.left_child]
+                    else:
+                        node = &self.nodes[node.right_child]
+                    current_depth += 1.
+                # Depth of node i = current_depth + average path length
+                depth_ptr[i] = <DOUBLE_t> current_depth + average_path_length(node.n_node_samples)
```

Comment:
Yes you are right. This should definitely be done at the Python level and have `_apply_depth_dense` only compute the actual depth. 

The point of the algorithm is to compute the depth of the leaf the sample would reach _if the tree was fully developed_, hence the added term.